Linting rule warning about using undefined translation strings

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

This is basically the flip side to https://github.com/learningequality/kolibri/pull/5790 - this linting rule will warn when we try to use a string that isn't defined.

Example:

```vue
<template>
    <div>
        <h1>{{ $tr('undefinedLabel') }}</h1>
    </div>
</template>

<script>
export default {
    name: 'my-component',
    $trs: {
        coachLabel: "Coach",
    }
}
</script>
```

In this case, there is an attempt to use a string by the key `undefinedLabel` - but it is not defined in $trs, so the linter will show an warning.

*Note that this will not address attempts to use common strings that are not defined.* So this works specifically at the level of an individual Vue component.

### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

Add this rule to the .eslintrc file to activate it. Add a call to `$tr()` somewhere in a component passing an undefined string to it. `yarn lint-frontend` and see the expected error.

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Supplementary to https://github.com/learningequality/kolibri/pull/5790

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
Inserted into linting est. structure

nucleogenesis committed:
Get uses and warn the string key and filepath

nucleogenesis committed:
Check for dynamic uses and don't warn for them

nucleogenesis committed:
Remove unnecessary comments

nucleogenesis committed:
Create eslint rule vue-no-unused-translations

nucleogenesis committed:
Remove unnecessary code

nucleogenesis committed:
Add tests

nucleogenesis committed:
Additional tests, dynamic uses, lint the linter

nucleogenesis committed:
Linting rule - warn on using undefined strings

jonboiser commented on packages/eslint-plugin-kolibri/lib/utils.js:
```diff
@@ -211,4 +211,33 @@ module.exports = {
+    badAttempts.forEach(node => {
+      context.report({
+        node,
+        message: `Attempt to use undefined string ${node.value}`,
```
Maybe just something more matter-of-fact like `Message not defined in $trs: "<string_id>"` and string_id also wrapped in quotes with printed out.

jonboiser commented on packages/eslint-plugin-kolibri/lib/utils.js:
```diff
@@ -211,4 +211,33 @@ module.exports = {
+    unused.forEach(prop => {
+      context.report({
+        node: prop.node,
+        message: `Unused translation defined for ${prop.node.name}`,
```
Have it match the style of the other errors, like `Unused message found in $trs: "<name>"`

jonboiser commented:
The 'unused' rule assumes that an ID appears as a value in `$trs`, it's being used. Screenshot shows no errors in my editor.

![Screen Shot 2019-08-07 at 2 13 13 PM](https://user-images.githubusercontent.com/10248067/62658423-a0887b80-b91d-11e9-93ae-cd4d394ed105.png)

jonboiser commented on packages/eslint-plugin-kolibri/lib/rules/vue-no-unused-translations.js:
```diff
@@ -0,0 +1,101 @@
+    {
+      ObjectExpression(node) {
+        node.properties.forEach(prop => {
+          if (prop.value.value) {
```
You need to make these guards a lot safer, since this rule was getting some 'undefined is not an object' errors on some of my files. I replaced all the guards with `lodash/get` and it seemed to fix it.

e.g.

```
if(get(value, ['value', 'value']) {...}
```

jonboiser commented on packages/eslint-plugin-kolibri/lib/rules/vue-no-unused-translations.js:
```diff
@@ -0,0 +1,101 @@
+    {
+      'CallExpression[callee.type="MemberExpression"]'(node) {
+        if (node.callee.property.name == $TR_FUNCTION && node.arguments.length) {
+          node.arguments.forEach(arg => usedStrings.push(arg.value));
```
Guard against `arg` being undefined here

jonboiser commented on packages/eslint-plugin-kolibri/lib/rules/vue-no-unused-translations.js:
```diff
@@ -0,0 +1,101 @@
+    {
+      'CallExpression[callee.type="Identifier"]'(node) {
+        if (node.callee.name == $TR_FUNCTION && node.arguments.length) {
+          node.arguments.forEach(arg => Boolean(arg.value) && usedStrings.push(arg.value));
```
Use an if-guard instead of short circuiting here

jonboiser commented:
In this PR, could you add the two new rules to `eslintrc`. We'll want to have them enabled by default.

jonboiser committed:
Require new string rule

jonboiser committed:
Make accessors on nodes safer

nucleogenesis committed:
Merge pull request #2 from jonboiser/lint-undefined-string-uses

Patches for  Lint undefined string uses

nucleogenesis commented:
Travis is failing because of the issues living in the code. Should we clean up to get the linter passing and keep it in eslintrc?

jonboiser commented:
Are there a lot of them? If it's not too many, we can put in eslint-disable-next-line comments in the mean time.

jonboiser commented:
There's some strings we don't want to delete because of crossComponentTranslator, but if you're sure that a certain clean up area is safe, let's go for it. (But handle the crossComponentTranslator case with the eslint-disable comments).

nucleogenesis commented:
> crossComponentTranslator

I previously removed all calls to crossComponentTranslator. The only instance of it is its definition. A part of the strings refactor (https://github.com/learningequality/kolibri/pull/5718) was to remove all uses of it, so any place where it was previously used to define strings - those strings were instead just defined on the component directly rather than using crossComponentTranslator.

> Are there a lot of them? If it's not too many, we can put in eslint-disable-next-line comments in the mean time.

There are about 30-40 individual instances across about a dozen files. 

In any case - if we're going to merge this with the rule enabled in `.eslintrc` then we ought to push a set of fixes in quick succession right?

The same goes for the flip-side of this PR (https://github.com/learningequality/kolibri/pull/5790) (which has far more instances that need to be fixed).

nucleogenesis committed:
Deactivate the rules to pass Travis

indirectlylit commented:
> In any case - if we're going to merge this with the rule enabled in .eslintrc then we ought to push a set of fixes in quick succession right?
> 
> The same goes for the flip-side of this PR (#5790) (which has far more instances that need to be fixed).

Yes, generally in the past we've combined the fixes and linter updates into a single PR (separate commits)

rtibbles commented:
In the interests of getting code more rapidly into develop, I merged #5790 without the corresponding fixes. I think it is reasonable to do the same here if this is good to go. The follow up PR can activate the linting and resolve the issues.

nucleogenesis committed:
Improve error message text

nucleogenesis commented:
I updated the error messages for both of the linting rules I added. With an approval it should be ready to merge. 

I will follow up with a PR applying fixes for the issues both rules found once merged.

rtibbles reviewed (approved):
Messaging comments addressed!

indirectlylit commented:
Looks like the tests are failing .. as intended :)

nucleogenesis committed:
Messages in spec as well
